### PR TITLE
Upstream ttrpc

### PR DESCRIPTION
**Pull request by lifupan**

### CODE REVIEW

**Comment by jodh-intel** [comment]
Hi @lifupan - thanks for raising this. I'll review later today. However, taking a look at ttRPC, from what I can see, there is no way to shut down the server. gRPC provides a `server.shutdown().wait()` API, but ttRPC doesn't appear to provide any shutdown API?. In fact, it doesn't even seem possible to `drop(server)` due to the way the ttRPC API has been written (`start()` passes ownership)?

The reason I'm raising this is that we need to be able to shutdown the server to allow agent tracing to work.

**Comment by lifupan** [comment]
> Hi @lifupan - thanks for raising this. I'll review later today. However, taking a look at ttRPC, from what I can see, there is no way to shut down the server. gRPC provides a `server.shutdown().wait()` API, but ttRPC doesn't appear to provide any shutdown API?. In fact, it doesn't even seem possible to `drop(server)` due to the way the ttRPC API has been written (`start()` passes ownership)?
> 
> The reason I'm raising this is that we need to be able to shutdown the server to allow agent tracing to work.

Hi @jodh-intel 

Yes, the ttrpc didn't support a shutdown api. BTW, why the trace needs a shutdown api? If it's must needed, I can add it.

**Comment by jodh-intel** [comment]
Hi @lifupan - tracing requires a shutdown because for trace "spans" (aka transactions) to complete, it *must* be possible for the entire agent process to shut down cleanly in a controlled fashion: stopping all threads, ending async logging correctly and then calling `std::process:exit()`.

I have a branch that handles all this for gRPC + openTelemetry but it's broken due to the old version of the `grpc-rs` crate we're using that supports vsock.

In summary: if we want tracing in the rust agent, we need a ttRPC shutdown mechanism :smile:

**Comment by lifupan** [comment]
> Hi @lifupan - tracing requires a shutdown because for trace "spans" (aka transactions) to complete, it _must_ be possible for the entire agent process to shut down cleanly in a controlled fashion: stopping all threads, ending async logging correctly and then calling `std::process:exit()`.
> 
> I have a branch that handles all this for gRPC + openTelemetry but it's broken due to the old version of the `grpc-rs` crate we're using that supports vsock.
> 
> In summary: if we want tracing in the rust agent, we need a ttRPC shutdown mechanism 😄

Hi @jodh-intel 
Got it, I'll add the suhtdown mechanism to ttrpc.

**Comment by jodh-intel** [comment]
Thanks very much @lifupan ! :smile:

**Comment by lifupan** [comment]
Hi, @jodh-intel 
I have two points needed to confirm with you about the ttrpc shutdown logic:

1) Does the shutdown function is a synchronized or asynchronous？ In another words, does the shutdown needed to wait all of the ttrpc serve threads exited or just notify ttrpc server to stop and return directly without waiting all of the threads exited? 

2) If it's synchronized, how could you make sure the ttrpc server shutdown cleanly before the sandbox shutdown the hypervisor? I noticed that once the agent completely finished the "destroy sandbox" rpc call and the sandbox would shutdown the VM but at the same time the agent in VM would hasn't finished shutdown rpc server cleanly.

**Comment by lifupan** [comment]
Hi @jodh-intel 

I had added a shutdown mechanism to ttrpc-rust, and you can have a try and see does it meet your needs.  https://github.com/containerd/ttrpc-rust/pull/24

**Comment by jodh-intel** [comment]
Hi @lifupan - regarding your questions:

----

1) The shutdown needs to be synchronous since if it isn't, there may be missed spans related to other (partially completed) ttrpc calls.

Here's a few pointers to how this works in the existing golang agent...

Here is the main threads gRPC handling code:

- https://github.com/kata-containers/agent/blob/master/agent.go#L1573..L1596

The main thread calls `waitForStopServer()` which waits for the gRPC server thread to close the channel and the main thread then stops the gRPC server by calling `server.GracefulStop()`. The main thread can then return to allow the agent process to exit cleanly.

Here is where the gRPC server thread requests that the main thread stop the gRPC server (in `DestroySandbox()`):

- https://github.com/kata-containers/agent/blob/master/grpc.go#L1530..L1534

----

2) Yes, getting all the components to work together is going to be difficult (it was for the existing golang agent tracing too ;)

If you look at the tracing document here:

- https://github.com/kata-containers/agent/blob/master/TRACING.md#trace-modes

What is *supposed* to happen is that when static agent tracing is configured in the runtime's `configuration.toml`, the *agent* shuts down the VM. Normally of course the runtime handles this.

**However**, it appears the static tracing codepath has broken. I'm looking at it, but I think the problem is due to the (relatively) new persist API in the runtime which is not storing the trace mode correctly, meaning the runtime will today kill the VM whatever trace mode you specify.

**Comment by jodh-intel** [comment]
@lifupan - thanks for adding the shutdown API to ttrpc. I've tested it, but I think I've got a deadlock - my ttrpc server thread has requested the shutdown, but the main thread is blocked on attempting to lock the connections mutex, presumably because the client call (`Server.shutdown()`) is still in progress:

- https://github.com/containerd/ttrpc-rust/blob/master/src/server.rs#L531

**Comment by lifupan** [comment]
Hi @jodh-intel 

Can you have a try with this patch applied on ttrpc-rust? 

```
diff --git a/src/server.rs b/src/server.rs
index fa1af12..517189a 100644
--- a/src/server.rs
+++ b/src/server.rs
@@ -521,6 +521,8 @@ impl Server {
     }
 
     pub fn shutdown(mut self) {
+        let connections = self.connections.lock().unwrap();
+
         self.quit.store(true, Ordering::SeqCst);
         close(self.monitor_fd.1).unwrap_or_else(|e| {
             warn!(
@@ -528,7 +530,6 @@ impl Server {
                 self.monitor_fd.1, e
             )
         });
-        let connections = self.connections.lock().unwrap();
 
         for (_fd, c) in connections.iter() {
             c.close();

```

**Comment by jodh-intel** [comment]
Hi @lifupan - that doesn't works (blocks), but this gets further:

```diff
diff --git a/src/server.rs b/src/server.rs
index fa1af12..7f68522 100644
--- a/src/server.rs
+++ b/src/server.rs
@@ -521,6 +521,9 @@ impl Server {
     }
 
     pub fn shutdown(mut self) {
+        let connections_ref = self.connections.clone();
+        let connections = connections_ref.lock().unwrap();
+
         self.quit.store(true, Ordering::SeqCst);
         close(self.monitor_fd.1).unwrap_or_else(|e| {
             warn!(
@@ -528,8 +531,6 @@ impl Server {
                 self.monitor_fd.1, e
             )
         });
-        let connections = self.connections.lock().unwrap();
-
         for (_fd, c) in connections.iter() {
             c.close();
         }
@@ -539,6 +540,7 @@ impl Server {
         drop(connections);
 
         if let Some(handler) = self.handler.take() {
+            // FIXME: blocks here
             handler.join().unwrap();
         }
     }
```

As shown, it blocks in `handler.join().unwrap()` though.

**Comment by lifupan** [comment]
Hi @jodh-intel 
Can you give me your case and let me reproduce it?

**Comment by jodh-intel** [comment]
Hi @lifupan - the source is here: https://github.com/jamesodhunt/test-protobuf-ttrpc (note you'll have to change `Cargo.toml` as I'm using a local git checkout for ttrpc).

**Comment by lifupan** [comment]
> Hi @lifupan - the source is here: https://github.com/jamesodhunt/test-protobuf-ttrpc (note you'll have to change `Cargo.toml` as I'm using a local git checkout for ttrpc).

Hi @jodh-intel  

I had fixed the lock issue and your test cases worked well from my side using rust-ttrpc:  https://github.com/lifupan/ttrpc-rust/tree/fix_deadlock 

You can have a try with a small fix to your codes with below, since you should close the connection fd in shutdown service, otherwise the shutdown service couldn't reply successfully and you client wouldn't get the answer and would wait for ever. 

```
diff --git a/src/server.rs b/src/server.rs
index 4e77df3..0c3f1c7 100644
--- a/src/server.rs
+++ b/src/server.rs
@@ -58,7 +58,7 @@ impl<'a> MyService for HelloService {
         info!(sl!(), "requesting shutdown");
 
         let result = tx.send(true);
-        let _ = close(ctx.fd);
+//        let _ = close(ctx.fd);
 
         info!(sl!(), "requested shutdown"; "result" => format!("{:?}", result));
 ```

**Comment by jodh-intel** [comment]
@lifupan - \o/ Thanks - yes, it now wfm! All the best fixes are "one liners" :smile:

**Comment by jodh-intel** [comment]
@lifupan - actually, I spoke too soon. It *does* work for me locally, but only if I remove https://github.com/lifupan/ttrpc-rust/commit/a54e9cfc996a752fec835135c9c945d3f8c2fe3c and add the fix I mentioned in https://github.com/kata-containers/kata-containers/pull/177#issuecomment-616525738. In summary, my diff from ttrpc's `master HEAD` (`b0194b38b1039748eed1ae674b9852ba6a84f985`) is:

```diff
diff --git a/src/server.rs b/src/server.rs
index f501416..05a0b5f 100644
--- a/src/server.rs
+++ b/src/server.rs
@@ -438,7 +438,6 @@ impl Server {
                     let quit = Arc::new(AtomicBool::new(false));
                     let child_quit = quit.clone();

-                    let connections_child = connections.clone();
                     let handler = thread::Builder::new()
                         .name("client_handler".into())
                         .spawn(move || {
@@ -490,8 +489,6 @@ impl Server {
                             handler.join().unwrap_or(());
                             close(fd).unwrap_or(());

-                            let _ = connections_child.lock().unwrap().remove(&fd);
-
                             info!("client thread quit");
                         })
                         .unwrap();
@@ -524,7 +521,8 @@ impl Server {
     }

     pub fn shutdown(mut self) {
-        let connections = self.connections.lock().unwrap();
+        let connections_ref = self.connections.clone();
+        let connections = connections_ref.lock().unwrap();

         self.quit.store(true, Ordering::SeqCst);
         close(self.monitor_fd.1).unwrap_or_else(|e| {
```

File: src/agent/netlink/Cargo.toml [hunk 1]
```diff
@@ -9,7 +9,7 @@ edition = "2018"
 [dependencies]
 libc = "0.2.58"
 nix = "0.17.0"
-protobuf = "2.6.1"
+protobuf = "2.8.1"
```
Comment (jodh-intel) [inline src/agent/netlink/Cargo.toml:12]: Whis is this version different to the top-level and `protocols` `Cargo.toml` (`2.14.0`)?

File: src/agent/src/rpc.rs [hunk 2]
```diff
@@ -521,150 +520,113 @@ impl agentService {
...
-        } else {
-            let resp = Empty::new();
-            let f = sink
-                .success(resp)
-                .map_err(move |_e| error!(sl!(), "fail to create container"));
-            ctx.spawn(f);
+    ) -> ttrpc::Result<Empty> {
+        match self.do_create_container(req) {
+            Err(e) => {
+                //error!(sl!(), format!("failed to create container: {}", e.to_string()));
```
Comment (jodh-intel) [inline src/agent/src/rpc.rs:531]: Nit: commented out code (here and below).

File: src/agent/src/rpc.rs [hunk 3]
```diff
@@ -1628,34 +1360,33 @@ fn find_process<'a>(
...
+    addr.push_str(":");
+    addr.push_str(&port.to_string());
+
+    let server = ttrpc::Server::new()
+        .bind(addr.as_str())
+        .unwrap()
+        .register_service(aservice)
+        .register_service(hservice);
+
+    info!(sl!(), "TTRPC server started");
```
Comment (jodh-intel) [inline src/agent/src/rpc.rs:1389]: OOI, is the official project name "ttrpc", "TTRPC" or "ttRPC"? I'd guess "ttRPC" as that seems to align most closely to "gRPC"?

**Comment by lifupan** [comment]
>                              close(fd).unwrap_or(());
> 
> -                            let _ = connections_child.lock().unwrap().remove(&fd);
> -

Hi @jodh-intel 

Yes, this change can make the shutdown work, but it will cause memory leak which didn't remove the connection from the hashmap  when it closed. That's what this https://github.com/containerd/ttrpc-rust/pull/32 try to fix.

File: src/agent/netlink/Cargo.toml (hunk 1 above)
Comment (lifupan) [reply src/agent/netlink/Cargo.toml:12]: yes, you are right, it should be "2.14.0".

File: src/agent/src/rpc.rs (hunk 3 above)
Comment (lifupan) [reply src/agent/src/rpc.rs:1389]: changed it to "ttRPC" .

**Comment by jodh-intel** [comment]
Hi @lifupan - confirmed - https://github.com/containerd/ttrpc-rust/commit/8ef0f3303b974bf07655decf9218f0137b19c34b now works for me \o/.

**Comment by lifupan** [comment]
> Hi @lifupan - confirmed - [containerd/ttrpc-rust@8ef0f33](https://github.com/containerd/ttrpc-rust/commit/8ef0f3303b974bf07655decf9218f0137b19c34b) now works for me \o/.

That's great.   
Hi  @yyyeerbo 
Can you take a look and if there is no other concerns then we can merge it first and then I can go on the runtime's ttrpc switch.
